[Link OTP in WalletsButton] Adds LinkEmbeddedInteractor to WalletsButton.

carlosmuvi-stripe opened the pull request:
# Summary
- Adds `LinkEmbeddedInteractor`, that will handle the inline OTP, Link payment method selection, etc.
- Hooks the interactor state into the `WalletButtonsState`.
- Adds a new Verification state section in the UI, and hides the wallet button while it's visible.

Note verification completion is not handled in this PR. 

https://github.com/user-attachments/assets/4c6d908c-1e45-4c49-b477-19206e759e82

# Motivation
<!-- Why are you making this change? If it's for fixing a bug, if possible, please include a code snippet or example project that demonstrates the issue. -->

# Testing
<!-- How was the code tested? Be as specific as possible. -->
- [ ] Added tests
- [ ] Modified tests
- [ ] Manually verified

# Screenshots
| Before  | After |
| ------------- | ------------- |
| *before screenshot*  | *after screenshot* |

# Changelog
<!-- Is this a notable change that affects users? If so, add a line to `CHANGELOG.md` and prefix the line with one of the following:
    - [Added] for new features.
    - [Changed] for changes in existing functionality.
    - [Deprecated] for soon-to-be removed features.
    - [Removed] for now removed features.
    - [Fixed] for any bug fixes.
    - [Security] in case of vulnerabilities.
-->

carlosmuvi-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/ui/wallet/LinkEmbeddedOtpSection.kt:
```diff
@@ -0,0 +1,117 @@
+import com.stripe.android.uicore.elements.OTPElementUI
+
+@Composable
+internal fun LinkEmbeddedOtpSection(
```
UI is not final yet. Once it is we'll add screenshot tests.

carlosmuvi-stripe commented on paymentsheet-example/src/main/java/com/stripe/android/paymentsheet/example/playground/PaymentSheetPlaygroundActivity.kt:
```diff
@@ -436,6 +438,7 @@ internal class PaymentSheetPlaygroundActivity :
             onClick = flowController::presentPaymentOptions
         )
         BuyButton(
+            modifier = Modifier.imePadding(),
```
On the playground, pushes the Flow controller area up with the keyboard.

carlosmuvi-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/verification/NoOpLinkEmbeddedInteractor.kt:
```diff
@@ -0,0 +1,22 @@
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+
+/**
+ * Implementation for surfaces that don't support Link embedded verification.
+ */
+internal class NoOpLinkEmbeddedInteractor : LinkEmbeddedInteractor {
```
Using this on the embedded integration since we're not planning on supporting WalletsButton there.

tillh-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/verification/LinkEmbeddedState.kt:
```diff
@@ -0,0 +1,39 @@
+/**
+ * Represents the different states of the verification process
+ */
+internal sealed class VerificationState : Parcelable {
```
The naming here threw me off a bit, especially `Verifying`. Should we rename this to better describe what the outcome is, like `Loading`, `RenderInlineVerification`, and `RenderButton`?

tillh-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/ui/wallet/LinkEmbeddedOtpSection.kt:
```diff
@@ -0,0 +1,117 @@
+                        enabled = !verificationState.isProcessing,
+                        element = otpElement,
+                        otpInputPlaceholder = " ",
+                        middleSpacing = 6.dp,
```
Can we set this to 8 dp, just like the `boxSpacing`?

tillh-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/ui/wallet/LinkEmbeddedOtpSection.kt:
```diff
@@ -0,0 +1,117 @@
+        textAlign = TextAlign.Center,
+        modifier = Modifier
+            .fillMaxWidth()
+            .padding(bottom = 16.dp)
```
Nit: Can we move this internal padding to the parent composable?

tillh-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/verification/DefaultLinkEmbeddedInteractor.kt:
```diff
@@ -0,0 +1,104 @@
+            )
+        )
+
+    override val state: StateFlow<LinkEmbeddedState> = _state
```
Nit: We don't really need the `_state` field.

tillh-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/ui/WalletButtonsContent.kt:
```diff
@@ -27,11 +31,22 @@ internal class WalletButtonsContent(
                 Column(
                     verticalArrangement = Arrangement.spacedBy(12.dp)
                 ) {
+                    state.linkOtpState?.let {
```
Oh, I see that's how it works thanks to the filtering in the interactor. Can you make this clearer with a comment?

samer-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/verification/LinkEmbeddedInteractor.kt:
```diff
@@ -0,0 +1,17 @@
+import com.stripe.android.uicore.elements.OTPElement
+import kotlinx.coroutines.flow.StateFlow
+
+internal interface LinkEmbeddedInteractor {
```
Should we rename this to something since it's no longer part of `Embedded`?

carlosmuvi-stripe commented:
@tillh-stripe @samer-stripe Addressed feedback!

- Renamed `LinkEmbeddedInteractor` to `LinkInlineInteractor` and added some docs.
- Cleaned up the UI and added previews (snapshots coming once we have final designs).
- Renamed verification states and added more clarifying docs.

tillh-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/verification/LinkInlineInteractor.kt:
```diff
@@ -4,9 +4,14 @@ import com.stripe.android.lpmfoundations.paymentmethod.PaymentMethodMetadata
+ *
+ * @see [com.stripe.android.paymentsheet.ui.WalletButtonsInteractor.WalletButton]
+ */
+internal interface LinkInlineInteractor {
```
Nit: Maybe we can even name it `LinkInlineVerificationInteractor` or `LinkInlineOtpInteractor`, to avoid confusion with Link inline signup.
